contracts/zaps/ZapTrancheHop.sol

```diff
@@ -0,0 +1,116 @@
+    bytes32 internal immutable _trancheBytecodeHash;
+
+    /// @param __trancheFactory Address of the TrancheFactory contract
+    /// @param __trancheBytecodeHash Hash of the Tranche bytecode.
+    constructor(address __trancheFactory, bytes32 __trancheBytecodeHash)
+        Authorizable()
+    {
+        _authorize(msg.sender);
+        _trancheFactory = __trancheFactory;
+        _trancheBytecodeHash = __trancheBytecodeHash;
+    }
+
+    /// @dev Requires that the contract is not frozen
+    modifier notFrozen() {
+        require(!isFrozen, "Contract frozen");
+        _;
+    }
+
+    /// @dev Allows an authorized address to freeze or unfreeze this contract
+    /// @param _newState True for frozen and false for unfrozen
+    function setIsFrozen(bool _newState) external onlyAuthorized() {
+        isFrozen = _newState;
+    }
+
+    /// @notice Redeems Principal and Yield tokens and deposits the underlying assets received into
+    /// a new tranche. The target tranche must use the same underlying asset.
+    /// @param _underlying The underlying ERC20 token contract of the yearn vault.
+    /// @param _positionFrom The wrapped position of the originating tranche.
+    /// @param _expirationFrom The expiration timestamp of the originating tranche.
+    /// @param _positionTo The wrapped position of the target tranche.
+    /// @param _expirationTo The expiration timestamp of the target tranche.
+    /// @param _amountPt Amount of principal tokens to redeem and deposit into the new tranche.
+    /// @param _amountYt Amount of yield tokens to redeem and deposit into the new tranche.
+    /// @param _ptExpected The minimum amount of principal tokens to mint.
+    /// @return returns the minted amounts of principal and yield tokens (PT and YT)
+    function hopToTranche(
+        IERC20 _underlying,
+        address _positionFrom,
+        uint256 _expirationFrom,
+        address _positionTo,
+        uint256 _expirationTo,
+        uint256 _amountPt,
+        uint256 _amountYt,
+        uint256 _ptExpected
+    ) public returns (uint256, uint256) {
+        ITranche trancheFrom = _deriveTranche(
+            address(_positionFrom),
+            _expirationFrom
+        );
+        ITranche trancheTo = _deriveTranche(
+            address(_positionTo),
+            _expirationTo
+        );
+
+        if (_amountPt > 0) {
+            trancheFrom.transferFrom(msg.sender, address(this), _amountPt);
+            trancheFrom.approve(address(trancheFrom), _amountPt);
+            trancheFrom.withdrawPrincipal(_amountPt, _positionTo);
+        }
+
+        if (_amountYt > 0) {
+            IERC20 yt = IERC20(trancheFrom.interestToken());
+            yt.transferFrom(msg.sender, address(this), _amountYt);
+            yt.approve(address(trancheFrom), _amountYt);
+            trancheFrom.withdrawInterest(_amountYt, _positionTo);
+        }
+
+        uint256 balance = _underlying.balanceOf(_positionTo);
```

Comment:
We can cut this call by using the return value of the withdraws

Reply:
❤️